src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -0,0 +1,215 @@
+        public GeneratorDriver WithPendingEdits(ImmutableArray<PendingEdit> edits)
+        {
+            var newState = _state.With(edits: _state.Edits.AddRange(edits));
+            return FromState(newState);
+        }
+
+        //PROTOTYPE
+        public bool ContainsAdditionalText(AdditionalText text) => _state.AdditionalTexts.Contains(text);
+
+        private static GeneratorDriverState ApplyPartialEdit(GeneratorDriverState state, PendingEdit edit, CancellationToken cancellationToken = default)
+        {
+            var initialState = state;
+
+            // see if any generators accept this particular edit
+            foreach (var provider in state.Providers)
+            {
+                var generator = provider.GetGenerator();
+                if (edit.AcceptedBy(generator))
+                {
+                    // attempt to apply the edit
+                    var context = new UpdateContext(state.Sources[provider], cancellationToken);
+                    var succeeded = edit.TryApply(generator, context);
+                    if (!succeeded)
+                    {
+                        // we couldn't successfully apply this edit
+                        // return the original state noting we failed
+                        return initialState.With(editsFailed: true);
+                    }
+
+                    // update the state with the new edits
+                    state = state.With(sources: state.Sources.SetItem(provider, context.AdditionalSources.ToImmutableAndFree()));
+                }
+            }
+
+            return state;
+        }
+
+        private static GeneratorDriverState StateWithPendingEditsApplied(GeneratorDriverState state)
+        {
+            if (state.Edits.Length == 0)
+            {
+                return state;
+            }
+
+            var newState = state;
+            foreach (var edit in newState.Edits)
+            {
+                newState = edit.Commit(newState);
+            }
+            return newState.With(edits: ImmutableArray<PendingEdit>.Empty, editsFailed: false);
+        }
+
+        private GeneratorDriver BuildFinalCompilation(Compilation compilation, out Compilation outputCompilation, GeneratorDriverState state, CancellationToken cancellationToken)
+        {
+            var finalCompilation = compilation;
+            foreach (var (_, sourcesAdded) in state.Sources)
+            {
+                foreach (var sourceText in sourcesAdded)
+                {
+                    try
+                    {
+                        //PROTOTYPE: should be possible to parallelize the parsing
+                        var tree = ParseGeneratedSourceText(sourceText, cancellationToken);
+                        finalCompilation = finalCompilation.AddSyntaxTrees(tree);
+                    }
+                    catch
+                    {
+                        //PROTOTYPE: should issue a diagnostic that the generator produced unparseable source
```

Comment:
The parser wouldn't throw though, right?

Reply:
That is a good point. We'll have to think about how we handle invalid code coming from the generator in the future. I guess we'll tag the syntax tree as being generated, then errors that come from it can be identified?

---
In reply to: [388647426](https://github.com/dotnet/roslyn/pull/41789#discussion_r388647426) [](ancestors = 388647426)

Reply:
So does the compiler do a special pass of outputting syntax errors before even producing the compilation? Is this running before or after that point?